src/unistd.rs

```diff
@@ -2778,3 +2778,42 @@ impl Group {
         })
     }
 }
+
+/// Get the name of the terminal device that is open on file descriptor fd
+/// (see [`ttyname(3)`](http://man7.org/linux/man-pages/man3/ttyname.3.html)).
+///
+/// # Safety
+///
+/// `ttyname()` mutates global variables and is *not* threadsafe.
+/// Mutating global variables is always considered `unsafe` by Rust and this
+/// function is marked as `unsafe` to reflect that.
+///
+/// For a threadsafe and non-`unsafe` alternative, see `ttyname_r()`.
+#[inline]
+pub unsafe fn ttyname(fd: RawFd) -> Result<String> {
+    let buf = libc::ttyname(fd);
+    if buf.is_null() {
+        return Err(Error::Sys(Errno::last()));
+    }
+
+    Ok(CStr::from_ptr(buf).to_string_lossy().into_owned())
+}
+
+/// Get the name of the terminal device that is open on file descriptor fd
+/// (see [`ttyname(3)`](http://man7.org/linux/man-pages/man3/ttyname.3.html)).
+///
+/// This is the threadsafe version of `ttyname()`.
+#[inline]
+pub fn ttyname_r(fd: RawFd) -> Result<String> {
+    const PATH_MAX: usize = libc::PATH_MAX as usize;
+    let mut buf = [0_u8; PATH_MAX];
+    let c_buf = buf.as_mut_ptr() as *mut libc::c_char;
+
+    let ret = unsafe { libc::ttyname_r(fd, c_buf, buf.len()) };
+    if ret != 0 {
+        return Err(Error::Sys(Errno::last()));
+    }
+
+    let nul = buf.iter().position(|c| *c == b'\0').unwrap();
```

Comment:
Instead of manually searching for NUL, why not just use `CStr::from_ptr`?  In fact, you should technically be using `OsString` here instead of `CString`, because the string comes from the operating system and not from another programming language.  So you should declare the buf as a `Vec`, and then use `OsString::from_vec` followed by `to_string_lossy`.

Reply:
i think returning a PathBuf is actually better than String here, and using OsString for that makes sense. that does mean that i do have to still search for the nul myself though, since OsString won't do that (it'll just create a string with a bunch of trailing nuls). going through CStr (and then being forced to allocate a copy into a second Vec) doesn't seem worth it to avoid that?